lib/time.c

```diff
@@ -23,22 +23,56 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 #include <apic.h>
+#include <errno.h>
+#include <percpu.h>
+#include <setup.h>
 #include <time.h>
 
-static volatile time_t ticks = 0;
+extern boot_flags_t boot_flags;
+
+static __aligned(16) volatile time_t ticks = 0;
 
 void timer_interrupt_handler(void) {
-    ++ticks;
+    asm volatile("lock incq %[ticks]" : [ ticks ] "=m"(ACCESS_ONCE(ticks)));
+    apic_EOI();
+}
+
+void apic_timer_interrupt_handler(void) {
+    asm volatile("lock incq %%gs:%[ticks]"
+                 : [ ticks ] "=m"(ACCESS_ONCE(PERCPU_VAR(apic_ticks))));
     apic_EOI();
 }
 
-void msleep(time_t ms) {
-    time_t end = ticks + ms;
-    while (ticks < end) {
+int msleep(time_t ms) {
+    time_t end;
+
+    if (!boot_flags.timer_global)
+        return -ENODEV;
+
+    end = ACCESS_ONCE(ticks) + ms;
+    while (ACCESS_ONCE(ticks) < end)
         cpu_relax();
-    }
+
+    return 0;
+}
+
+int msleep_local(time_t ms) {
+    time_t end;
+
+    if (!PERCPU_GET(apic_timer_enabled))
+        return -ENODEV;
+
+    end = PERCPU_GET(apic_ticks) + ms;
+    while (PERCPU_GET(apic_ticks) < end)
+        cpu_relax();
+
+    return 0;
 }
 
 time_t get_timer_ticks(void) {
-    return ticks;
+    return ACCESS_ONCE(ticks);
```

Comment:
is access once really necessary if `ticks` is defined as volatile already?

Reply:
No, it is not necessary. I put it nevertheless just in case the `volatile` disappears in the future and to highlight the atomic nature of the variable.

Reply:
it's not really a big deal i suppose.. It just makes it slightly unclear whether to use `ACCESS_ONCE(PERCPU_VAR(apic_ticks))` or `PERCPU_VAR(apic_ticks)`. The patch appears to use both...